docs/rfcs/0003-rfc-controller_refactor.md

```diff
@@ -0,0 +1,111 @@
+- Subscription status: ensure that user has an active subscription, and if not, show the "No Subscription Found" screen.
+
+The logic to calculate elapsed time since the VPN was activated will also be moved out of the Controller and into a new `Timer` component. This bit of work will likely require a separate RFC. 
+
+
+The `ConnectionManager` will have two state machines:
+1. `VPNClientState` to determine if the VPN client is off, connecting, or on. 
+2. `ConnectionState` to keep track of which connection related task the VPN is performing at any given point. 
+
+Before the VPN is toggled on, it is in `VPNClientStateOff`. When user toggles on the VPN, the client goes into `VPNClientStateConnecting`, the connection manager will perform the needed connection checks such as `ConnectionStateInternetProbe` and `ConnectionStateServerProbe`. If they all succeed, the VPN will enter `VPNClientStateOn` and call `Controller::activate()`.
+
+The `ConnectionManager` will communicate to the `Controller` to `activate()` and `deactivate()` the VPN. During one of the iterations of this proposal, someone asked if we can have signals instead of states in the `ConnectionManager`. The short answer is no. While emitting a signal to notify the UI about what we should be showing the user does the same thing, we still need to keep track of the current state internally so we know what to do next in case anything goes wrong at any other point.
+
+Here is the proposed set of `ConnectionManager` states and how they map to the `Controller` states. Please note that some of these states may go away once we are further along the implementation (e.g. we may not need `ConnectionStateSilentSwitching` anymore):
+
+```c++
+  enum VPNClientState {
+    VPNClientStateOff,          // VPN is off
+    VPNClientStateConnecting,   // VPN is perform connection and 
+                                // subscription probes in ConnectionState
+    VPNClientStateOn,           // VPN is on
+  }
+
+  enum ConnectionState {
+    ConnectionStateOff,                // Corresponds to client state:
+                                       // VPNClientStateOff
+    ConnectionStateInternetProbe,      // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateServerProbe,        // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateFirewall,           // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateCaptivePortal,      // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateCheckSubscription,  // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateOn,                 // Corresponds to client state:
+                                       // VPNClientStateOn
+  };
+```
```

Comment:
Above there is some mention of keeping the Unstable state, but it's not listed here.

Reply:
Above you also state that ConnectionState exists to "keep track of which connection related task the VPN is performing at any given point". "Off" and "On" are not really "tasks", as I understand it. Why are they part of this state?

Reply:
> Above you also state that ConnectionState exists to "keep track of which connection related task the VPN is performing at any given point". "Off" and "On" are not really "tasks", as I understand it. Why are they part of this state?

I have those connections in `ConnectionState `to map to when `VPNClientState` is either on of off because otherwise when the `VPNClientState` is On or Off we weren't very clean on what state `ConnectionState` would be in